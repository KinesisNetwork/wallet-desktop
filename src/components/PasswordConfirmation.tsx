import { focus } from '@helpers/focus'
import { decryptWithPassword } from '@services/encryption'
import { Wallet } from '@types'
import * as React from 'react'
import * as ReactDOM from 'react-dom'
import * as SwalImport from 'sweetalert'

const swal = SwalImport as any

export class PasswordInput extends React.Component<{ activeWallet: Wallet }, { password: string }> {
  displayName = 'PasswordInput'

  constructor(props) {
    super(props)
    this.state = {
      password: '',
    }
  }

  changeText = e => {
    const text = e.target.value
    this.setState({ password: text })
    const decryptedPrivateKeyOrEmpty = this.decryptKey(text)
    // @ts-ignore
    if (decryptedPrivateKeyOrEmpty !== '') {
      // Returns the decrypted private key
      swal.setActionValue({ cancel: { value: decryptedPrivateKeyOrEmpty } })
      swal.close('cancel')
    }
  }

  decryptKey = (passwordInput: string) => {
    return decryptWithPassword(this.props.activeWallet.encryptedPrivateKey, passwordInput)
  }

  render() {
    setTimeout(() => focus('popup-password-confirmation'))
    return (
      <input
        id="popup-password-confirmation"
        type="password"
        className="input is-large has-text-centered"
        value={this.state.password}
        onChange={ev => this.changeText(ev)}
      />
    )
  }
}

export const getPasswordConfirmation = async (
  activeWallet: Wallet,
  mode = 'danger',
): Promise<{ value: string }> => {
  const wrapper = document.createElement('div')
  ReactDOM.render(<PasswordInput activeWallet={activeWallet} />, wrapper)
  const element = wrapper.firstChild as Node

  const typeOfPopup = mode === 'danger' ? { icon: 'warning', dangerMode: true } : { icon: mode }
  return await sweetAlert({
    buttons: {
      cancel: {
        value: '',
        visible: false,
      },
    },
    content: { element },
    text: 'Please input password to confirm',
    ...typeOfPopup,
  })
}
