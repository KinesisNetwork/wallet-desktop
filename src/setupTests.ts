import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
