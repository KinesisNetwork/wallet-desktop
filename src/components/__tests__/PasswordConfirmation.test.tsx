import { shallow } from 'enzyme';
import * as React from 'react'

import { PasswordInput } from '@components/PasswordConfirmation'
import '../../setupTests'

describe('PasswordInput', () => {
  let props

  beforeEach(() => {
    props = {
      activeWallet: {
        accountName: 'myAccount',
        encryptedPrivateKey: '123',
        publicKey: 'qwe'
      }
    }
  })

  it('should render correctly', () => {
    const wrapper = shallow(<PasswordInput {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('returns an input field with the entereed password', () => {
    const wrapper = shallow(<PasswordInput {...props} />)

    expect(wrapper.find('input')).toHaveLength(1)

    wrapper.setState({ password: 'asd' })

    expect(wrapper.find('input').prop('value')).toEqual('asd')
  })

  it('triggers the changeText method on entering values in the input field', () => {
    const wrapper = shallow(<PasswordInput {...props} />)
    const changeTextSpy = jest
      .spyOn(wrapper.instance() as PasswordInput, 'changeText')
      .mockImplementation(() => null)
    const eventObj = { target: { value: 'a' } }

    wrapper.find('input').simulate('change', eventObj)

    expect(changeTextSpy).toHaveBeenCalledWith(eventObj)

    changeTextSpy.mockReset()
  })
})
