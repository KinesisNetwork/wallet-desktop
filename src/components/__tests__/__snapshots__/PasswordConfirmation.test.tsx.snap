// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput should render correctly 1`] = `
<input
  className="input is-large has-text-centered"
  id="popup-password-confirmation"
  onChange={[Function]}
  type="password"
  value=""
/>
`;
